### ISSUE TITLE
Consider using pinnable heap for AsymmetricAlgorithm.ExportParameters

### ISSUE THREAD

**Issue by filipnavara** [dotnet/runtime#46498]
.NET 5 introduced the concept of special heap for objects that are going to be pinned in memory. Using the [GC.AllocateArray API](https://docs.microsoft.com/en-us/dotnet/api/system.gc.allocatearray?view=net-5.0) new byte arrays could be placed into this heap. This has the additional guarantee that the memory won't be moved by the garbage collector and leave behind a copy of the data which is a valuable property for sensitive data. Various implementations of asymmetric cryptography algorithms (RSA, DSA, ECDiffieHellman, ECDsa) offer the `ExportParameters` method that includes sensitive private key data in a byte array. Using the pinnable heap for the private key byte array could allow the caller to securely dispose of the private key data once it is no longer used without worrying about the GC taking an accidental copy of the data. I am therefore proposing to modify the existing implementations to use the pinnable heap for the sensitive data to allow safer code like this:
```csharp
var dsaParameters = dsa.ExportParameters(true);
// Do something with the private key, like export to file
CryptographyOperations.ZeroMemory(dsaParamaters.X);
```

/cc @bartonjs @vcsjones

**Comment by Wraith2** [comment]
Does pinned memory allocated through that api get unpinned by the GC if it is found to be eligible for collection? If not there would need to be a disposable wrapper around it to encourage users to use it in a scoped manner. I know users _shouldn't_ be dropping things like private key material to the GC but you know that somewhere someone will be.

**Comment by filipnavara** [comment]
I assume it garbage collectible like a normal array would be.

Changing the existing `ExportParameters` API is unfortunately not an option but even the internal usage suffers from a small time window where the arrays can get moved by the GC.

**Comment by Wraith2** [comment]
Internal buffers sure, they should be changeable without external observation. It'd be nice to get confirmation on POH allocated objects and GC behaviour before considering exposing them to users.

**Comment by vcsjones** [comment]
Maybe @GrabYourPitchforks has some thought / input on this.

I'm not sure the impact or desire to return things from the POH every time sensitive data might be returned.

Perhaps it might be more useful to accept a Span to write to. That way we could write to a non movable buffer if that's where you want it written.

**Comment by filipnavara** [comment]
> Perhaps it might be more useful to accept a Span to write to. That way we could write to a non movable buffer if that's where you want it written.

That's what the newer APIs do. The problem is that due to the inheritance model of the classes they all end up using `ExportParameters` under the hood since different implementations can override it. The new APIs wrap the secret data in `fixed` block as soon as it's returned from `ExportParameters` which still leaves a small window of opportunity where the data can be moved by the GC.

**Comment by bartonjs** [comment]
Absent any gotchas, it seems like a reasonable change to me.

**Comment by Maoni0** [comment]
> Does pinned memory allocated through that api get unpinned by the GC if it is found to be eligible for collection?

yes - if an object is eligible for collection it means it will be gone (at some point). think an object on the POH same as any other object, it's just guaranteed to not be moved as long as it's alive.

**Comment by Wraith2** [comment]
Excellent, thank you.

**Comment by vcsjones** [comment]
> Absent any gotchas, it seems like a reasonable change to me.

I thought about this and one concern I have is that things in the POH are automatically on Gen2 of the GC. Lots of small, pinned, allocations might cause some interesting behaviors with the GC.

**Comment by vcsjones** [comment]
I am going to close this out because the [current guidance](https://github.com/dotnet/runtime/pull/97447#discussion_r1464923120) is that `GC.AllocateUninitializedArray<byte>(..., pinned: true);` should not be used for short-lived objects, which `ExportParameters`'s output is probably going to fit in to. This would fragment the POH and hurt the performance of applications, considering that the runtime will use `ExportParameters` itself sometimes.

We removed our use of the POH in crypto here: https://github.com/dotnet/runtime/pull/99168
